extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -0,0 +1,1212 @@
+
+\subsection{Structure of the document}
+\label{sec:doc-structure}
+This document proposes three extensions of the basic UTXO model, each
+one more general than the previous one (XUTXO \blue{(or EUTXO or whatever)} stands for
+\textit{Extended UTXO}):
+
+\begin{itemize}
+  \item \textbf{XUTXO-1} (Section~\ref{sec:xutxo-1}): this extends the
+    basic UTXO model with enhanced scripting features, allowing the
+    implementation of complex smart contracts.
+  \item \textbf{XUTXO-2} (Section~\ref{sec:xutxo-2}): this adds
+    multicurrency features to XUTXO-1, allowing users to define
+    \textit{custom currencies}.
+  \item \textbf{XUTXO-3} (Section~\ref{sec:xutxo-3}): this adds direct
+    support for non-fungible tokens to XUTXO-2.
+\end{itemize}
+
+\noindent We give formal specifications of each of these versions,
+based on the models appearing in the
+papers~\cite{Zahnentferner18-Chimeric} and
+\cite{Zahnentferner18-UTxO}.
+
+\medskip
+
+The rationale for providing three separate specifications is that (1)
+introducing extensions one by one clarifies the structure of the
+models and makes it easier to explain the relevant design decisions,
+and (2) it is possible that a particular blockchain might not need the
+full power of XUTXO-3 and so could use a simpler model, perhaps with
+less computational overhead.
+
+\bigskip
+
+\noindent Before presenting the extended models we will define some
+basic notation and give a little background on the standard UTXO
+model.
+
+
+\section{Notation}
+This section defines some basic notation.  We generally follow the
+notation established by \cite{Zahnentferner18-UTxO}.
+
+\begin{ruledfigure}{H}
+  \begin{itemize}
+\item Types are typeset in $\mathsf{sans~serif}$.
+
+\item All scripts are of type $\mathsf{Script}$.
+
+\item The only operation on $\mathsf{Script}$ from the ledger's
+  perspective is applying a validator script to three arguments,
+  which is denoted by $\llbracket \cdots \rrbracket :
+  \mathsf{Script} \rightarrow \mathsf{Script} \times \mathsf{Script}
+    \times \mathsf{Script} \rightarrow \mathbb{V}$ where $\mathbb{V}$
+    is some type containing a distinguished value
+    \textsf{error}\footnotemark.
+
+\item A record type with fields $\phi_1, \ldots, \phi_n$ of types $T_1,
+  \ldots, T_n$ is denoted by $(\phi_1 : T_1, \ldots, \phi_n : T_n)$.
+
+\item If $t$ is a value of a record type $T$ and $\phi$ is the name
+  of a field of $T$ then $t.\phi$ denotes the value of $\phi$ for
+  $t$.
+  
+\item A list $\lambda$ of type $\List{T}$ is
+    either the empty list $[]$ or a list $e :: \lambda'$ with $head$
+    $e$ of type $T$ and $tail$ $\lambda'$ of type
+    $\List{T}$. We denote the $i$th element of a list
+    $L$ by $L[i]$.
+    The concatenation of two lists $\lambda_1$ and
+    $\lambda_2$ is denoted $\lambda_1 ::: \lambda_2$.
+
+  \item For two types $K$ and $V$, $\Map{K}{V}$ denotes the type of
+    partial functions from \textit{keys} in $K$ to \textit{values} in
```

Comment:
I don't think that matters.  The thing which we have to avoid is adding up infinitely many values, and that can't happen if we only have a finite number of inputs and outputs.  It doesn't explicitly require that anywhere, but I'll fix it so it does.